messaging/src/main/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventHandlerInvoker.java

```diff
@@ -116,33 +126,75 @@ public void handle(@Nonnull EventMessage<?> message, @Nonnull Segment segment) t
             }
-            try {
-                super.invokeHandlers(message);
-            } catch (Exception e) {
-                DeadLetter<EventMessage<?>> letter = new GenericDeadLetter<>(sequenceIdentifier, message, e);
-                EnqueueDecision<EventMessage<?>> decision = enqueuePolicy.decide(letter, e);
-                if (decision.shouldEnqueue()) {
-                    Throwable cause = decision.enqueueCause().orElse(null);
-                    queue.enqueue(sequenceIdentifier, decision.withDiagnostics(letter.withCause(cause)));
-                } else if (logger.isInfoEnabled()) {
-                    logger.info("The enqueue policy decided not to dead letter event [{}].", message.getIdentifier());
-                }
+            invokeHandlers(message, segment, sequenceIdentifier);
+        }
+    }
+
+    private void invokeHandlers(@Nonnull EventMessage<?> message, @Nonnull Segment segment, Object sequenceIdentifier) {
+        if (logger.isTraceEnabled()) {
+            logger.trace("Event [{}] with queue id [{}] is not present in the dead-letter queue."
+                                 + "Handle operation is delegated to the wrapped EventHandlerInvoker.",
+                         message, sequenceIdentifier);
+        }
+        try {
+            super.invokeHandlers(message);
+        } catch (Exception e) {
+            DeadLetter<EventMessage<?>> letter = new GenericDeadLetter<>(sequenceIdentifier, message, e);
+            EnqueueDecision<EventMessage<?>> decision = enqueuePolicy.decide(letter, e);
+            if (decision.shouldEnqueue()) {
+                Throwable cause = decision.enqueueCause().orElse(null);
+                markPresentInDLQ(sequenceIdentifier, segment);
+                queue.enqueue(sequenceIdentifier, decision.withDiagnostics(letter.withCause(cause)));
+            } else if (logger.isInfoEnabled()) {
+                logger.info("The enqueue policy decided not to dead letter event [{}].", message.getIdentifier());
+            }
+        }
+    }
+
+    private boolean skipIfPresentCheck(
+            @Nonnull EventMessage<?> message,
+            @Nonnull Object sequenceIdentifier,
+            @Nonnull Segment segment
+    ) {
+        if (message instanceof DomainEventMessage) {
+            DomainEventMessage<?> domainEventMessage = (DomainEventMessage<?>) message;
+            if (domainEventMessage.getSequenceNumber() == 0L) {
+                markNotPresentInDLQ(sequenceIdentifier, segment);
+                return true;
             }
         }
+        if (!cacheEnabled) {
+            return false;
+        }
+        return cache.computeIfAbsent(segment.getSegmentId(), () ->
+                            new DeadLetteringCacheEntry(segment.getSegmentId(), cacheSize, queue))
+                    .skipIfPresentCheck(sequenceIdentifier);
+    }
+
+    private void markPresentInDLQ(@Nonnull Object sequenceIdentifier, @Nonnull Segment segment) {
```

Comment:
I think `markEnqueued` works better for this method name.

Reply:
We can simply check on the boolean flag like it was before. Renamed.